### DISCUSSION TITLE
Clarity on .NET Standard, Base Class Libraries, Runtime and .NET 5?

### DISCUSSION THREAD

**Discussion by BandaruNithin** [dotnet/runtime#50729] [General]
Hi,

I have come across this below the image in this URL: [https://github.com/dotnet/designs/blob/main/accepted/2020/net5/net5.md](https://github.com/dotnet/designs/blob/main/accepted/2020/net5/net5.md).

![image](https://user-images.githubusercontent.com/43375874/113404697-319b4880-93c6-11eb-974a-5cf34f4e212e.png)

**This is my understanding until now:**

**.NET Standard and Base Class Library:**
1. .NET Standard will provide principles and definitions of the .NET Standard. Code implementation is not written in .NET Standard [https://github.com/dotnet/standard](https://github.com/dotnet/standard)
**Q1. Is my assumption right as I did not find any code in this repository and found only abstracts of them?**

2. In many articles it's said that .NET Standard is introduced to avoid fragmentation and can share the same codebase by different .NET platforms which didn't give much clarity on how things changed at the repositories level.
**Q2. Is it like in the beginning all platforms of .NET like .NET framework, Xamarin, .net core had different functionalities based on their needs?**
**Q3. Then .NET standard library was introduced and defined all these basic components and asked all these platforms to implement .NET standard library principles and definitions. Is this correct?**
**Q4. Is Base Class Library(Core Library in .NET core) block in the above image represents the implementation of .NET Standard or its different?**

3. Based on Q4's answer I have further doubts.
**Q5. Is CoreFX repository the Base Class Library of .NET Core before or its different?**
**Q6. Now in .NET 5 instead of having multiple implementations, there will be only one code base for all. Is this correct?**

4. Now I can find the code base for System.Collections.Generic, System.Data etc in [https://github.com/dotnet/runtime](https://github.com/dotnet/runtime) repository which will produce a package called Microsoft.NetCore.App as mentioned in the following link [https://github.com/dotnet/announcements/issues/119](https://github.com/dotnet/announcements/issues/119).
**Q7. But when I look into the DLL's folders I can find these DLL's in both NETStandard.Library and Microsoft.NetCore.App folders in C:\Program Files\dotnet\packs. So, is there a separate repository for NETStandard.Library package or is it produced just based on build configuration?**

5. In both NETStandard.Library and Microsoft.NetCore.App I found Microsoft.Win32.Primitives.dll, System.dll, etc which have classes like Dictionary which will be platform dependent as it is only for windows if I am not wrong. 
**Q8. Is it not only windows DLL, if so then why is it part of the NETStandard.Library and Microsoft.NetCore.App as both needs to be portable and cross-platform or it's cross-platform because it has implementations for multiple platforms like using if-else or platform support attributes etc to behave based on platform type?**

**JIT and Runtime**

Now in .NET 5 is platform-specific code from Mono and .NET framework is being moved into .NET Core to make .NET Core(CoreCLR) support all platforms being a single code base.
**Q9. Is my assumption correct?**
**Q10. Is this being achieved using if-else conditions and platform support attribute to behave based on platform type?**

**General:**

How is code for different versions maintained in the repository? Ideally, one branch for each .NET version released and one that is being developed.
**Q11. If my assumption is right can you please provide some information about it?**

**Appended**

Comparing **C:\Program Files\dotnet\packs\NETStandard.Library.Ref\2.1.0\ref\netstandard2.1** and **C:\Program Files\dotnet\packs\Microsoft.NETCore.App.Ref\5.0.0\ref\net5.0** is net5.0 has many DLL's(251) compared to netstandard2.1(122). This is probably because .NET Standard is built long back and is not being updated now and .NET core is being updated instead of .NET Standard and from now on .NET Standard is continued as .NET 5.0 BCL.
**Q12. Is the above assumption correct?**

**NOTE:** _Maybe many of my questions would have been answered online somewhere but I am currently in a confused state and not sure if my understandings are right, so wanted clarity here._

**Comment by huoyaoyuan** [comment]
A1: Yes, if you are talking about the dotnet/standard library.

A2: Xamarin/mono was originally third party and only implemented part of them. .NET Core/.NET for Windows Store was cut from .NET Framework and dropped some functionalities.

A3: For versions before 2.0, the standard was defined by finding the common intersects the platform and interfaces have implemented, in order to unify them. 2.1 was asked to be implemented by actively developed ones(.NET Core and Mono)

A4: It's the implementation of all first party APIs, more than .NET Standard.

A5: Yes

A6: .NET 5 consolidates two things:
coreclr and corefx repos are merged, simplifying versioning and cross-repo changes.
Mono starts to share the BCL implementation(corefx)

A7: .NET standard is built from dotnet/standard repo. I remember there was it's building infrastructure in runtime repo, but removed.

A8: I'm not sure what you are referring to, but you are probably wrong.
.NET Standard is an interface which is not associated to any implementation.
The System.Private.CoreLib.dll or mscorlib.dll implementation depends on clr.dll/coreclr.dll, which is platform.specific.

A9 A10: .NET Core was forked from .NET Framework, then involved independently. The codebase contains abstraction layer (like #define) for different OSs, and the majority of code is platform.independent.

A11: Right. When time gets closed to release, a new branch is created for the release and patches will be commited onto it.

**Reply by BandaruNithin** [reply]
Can I get more clarity on **A7**? [https://github.com/dotnet/standard](https://github.com/dotnet/standard) doesn't have any implementation code, it has only an abstract of it. So, it is not possible to generate a DLL from it right?

One more thing I observed comparing **C:\Program Files\dotnet\packs\NETStandard.Library.Ref\2.1.0\ref\netstandard2.1** and **C:\Program Files\dotnet\packs\Microsoft.NETCore.App.Ref\5.0.0\ref\net5.0** is net5.0 has many DLL's(251) compared to netstandard2.1(122). This is probably because .NET Standard is built long back and is not being updated now and .NET core is being updated instead of .NET Standard and from now on .NET Standard is continued as .NET 5.0 BCL.
**Q12. Is the above assumption correct?**

**Regarding Answer A8:**
I am referring to these two folders **C:\Program Files\dotnet\packs\NETStandard.Library.Ref\2.1.0\ref\netstandard2.1** and **C:\Program Files\dotnet\packs\Microsoft.NETCore.App.Ref\5.0.0\ref\net5.0**
Here as you said NETStandard.Library and Microsoft.NETCore.App both has mscorlib.dll.

**Q. If this DLL is platform-specific how can it be part of NETStandard.Library or Microsoft.NETCore.App as these both need to be cross-platform and should be platform-independent(Should work for all platforms)?**

**Note:** _Added the above question to main at the end as Q12 to keep additional questions consolidated._

**Reply by huoyaoyuan** [reply]
> So, it is not possible to generate a DLL from it right?

A7: No. Ref assembly can be used to generate dll. If you decompile the netstandard.dll **in NETStandard.Library**, you can see it contains only `throw null`s.

A12: Not pretty. They are designed in different minds. The name `netstandard.dll` was picked *to not be conflict/confused with any existing dll*.

A8: NETStandard.Library and Microsoft.NETCore.App.Ref are all reference assemblies and contains no implementation. A platform implements .NET Standard by adding a `netstandard.dll` with implement forwarding in it's implementation pack, which is Microsoft.NETCore.App. The API surface is all the same in all platforms. **The API isn't associated to only one platform.**

**Reply by huoyaoyuan** [reply]
Demonstrate how .NET Standard works:
When a library is compiled against netstandard2.0, it doesn't copy `netstandard.dll` into its output. It just records "I need [netstandard]System.Object".
When a .NET Framework application loads this library, the CLR goes to use *its netstardard.dll*. The netstardard.dll in .NET Framework says "[netstandard]System.Object is forwarded to (being an alias of) [mscorlib]System.Object". Then the CLR goes to load from its mscorlib, which is platform dependent.
Similarly, netstardard.dll in .NET Core says "[netstandard]System.Object is forwarded to (being an alias of) [System.Private.CoreLib]System.Object".
